### ISSUE TITLE
[css-color-4] What if gradients with legacy colors *also* interpolated in Oklab by default?

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7948]
Right now, [css-color-4](https://drafts.csswg.org/css-color-4/#interpolation-space) allows UAs to handle interpolation between legacy colors in gamma-encoded sRGB: 

> However, user agents _may_ handle interpolation between legacy sRGB color formats (hex colors, named colors, [rgb()](https://drafts.csswg.org/css-color-4/#funcdef-rgb), [hsl()](https://drafts.csswg.org/css-color-4/#funcdef-hsl) or [hwb()](https://drafts.csswg.org/css-color-4/#funcdef-hwb) and the equivalent alpha-including forms) in gamma-encoded sRGB space. This provides Web compatibility; legacy sRGB content interpolates in the sRGB space by default.

I wrote this as a *may* to give UAs the choice of not doing this, but so far implementations seem to have all done this.
We have accepted that we have to live with this wart because of webcompat, but I'm not sure we have sufficiently examined the implications of the opposite. Can we actually examine the web compat implications of doing away with this clause? What if we moved all color interpolation on the Web away from gamma-encoded sRGB? 

There is precedent of changing rendering of existing content when the new way is universally better (`text-decoration-skip`, I'm looking at you), so this shouldn't be a non-starter, the bar is just high for what kind of changes are acceptable, and I'd argue this may pass that bar. Interpolating colors in Oklab never produces worse results (I challenge anyone who disagrees to find me an example where gamma-encoded sRGB interpolation produces better results, [you can play with the colors here in Safari TP](https://codepen.io/leaverou/pen/dyKbavR)).

The only case I can think of where gamma-encoded sRGB interpolation is desirable is color pickers. However, 
- the gradients in color pickers are only a visual aid, but the actual picked/displayed color would not be affected 
- I *think* this would only affect RGB color pickers ([like this](https://play.google.com/store/apps/details?id=com.edugameapp.rgbcolorpicker&hl=en_US&gl=US)), and most color pickers use a hue-based model (HSL, HSV, etc). I really struggled to find one that actually shows RGB visualizations. The hue-based ones are still implemented through segments of gamma-encoded sRGB interpolated gradients (since that was the only gradient available), but the segments are so small I don't believe there will be any difference.
- There should be wording in that spec that forbids UAs from implementing this without implementing the syntax to control interpolation space, so that authors can always revert to the previous behavior for the handful of cases where gamma-encoded sRGB interpolation is desirable.

It would be nice if gradients could just do the right thing by default without people needing to know about color spaces, and having to type `in oklab` in their gradients like a magic incantation until the end of time. 😕

## Examples

`red` to `lime`:
<img width="866" alt="image" src="https://user-images.githubusercontent.com/175836/197842652-006d3f6c-0e9d-4232-92c9-e0ee08df0d89.png">

`red` to `aqua`:
<img width="867" alt="image" src="https://user-images.githubusercontent.com/175836/197842798-ef9f050a-a0d5-47c6-b760-b4358f5f8a02.png">

`red` to `fuchsia`:
<img width="864" alt="image" src="https://user-images.githubusercontent.com/175836/197842882-d09e5dbc-fc2f-4f4f-94b1-530e487811d5.png">

`aqua` to `fuchsia`:
<img width="865" alt="image" src="https://user-images.githubusercontent.com/175836/197843062-54bf6124-1b05-4df1-aae9-ea945a529e72.png">

`lime` to `fuchsia`:
<img width="868" alt="image" src="https://user-images.githubusercontent.com/175836/197843159-5e91d938-31b5-44ef-86ff-441e35917bc2.png">

`lime` to `blue`:
<img width="868" alt="image" src="https://user-images.githubusercontent.com/175836/197843238-6a02196e-f8c8-4abf-8c5b-77abd7e2f0d6.png">

`blue` to `gray`:
<img width="864" alt="image" src="https://user-images.githubusercontent.com/175836/197843360-9c50a4c7-8392-4106-abaf-dd9f2a3b3e73.png">

`white` to `blue`:
<img width="858" alt="image" src="https://user-images.githubusercontent.com/175836/197843624-df45b396-9ff5-4078-ad8f-95c4e4fcb64e.png">

`white` to `black`:
<img width="862" alt="image" src="https://user-images.githubusercontent.com/175836/197843444-e864aa1c-e013-44b3-994e-57e9429e53c4.png">

**Comment by svgeesus** [comment]
> (I challenge anyone who disagrees to find me an example where gamma-encoded sRGB interpolation produces better results, [you can play with the colors here in Safari TP](https://codepen.io/leaverou/pen/dyKbavR))

Thanks for the demo! I had expected it to also work in Chrome Canary 108 (with experimental flag), but it didn't which I found surprising.

**Comment by LeaVerou** [comment]
> > (I challenge anyone who disagrees to find me an example where gamma-encoded sRGB interpolation produces better results, [you can play with the colors here in Safari TP](https://codepen.io/leaverou/pen/dyKbavR))
> 
> Thanks for the demo! I had expected it to also work in Chrome Canary 108 (with experimental flag), but it didn't which I found surprising.

It appears to be parsing the syntax, but not actually applying it. I guess that must be temporary, otherwise it would break feature detection! @sesse could clarify.

**Comment by romainmenke** [comment]
I think this is very risky.

> (I challenge anyone who disagrees to find me an example where gamma-encoded sRGB interpolation produces better results,

I don't think that is the right question.
All current usage should be assumed to be deliberate, matching some design/intention.

A preprocessor like PostCSS easily allows you to make `in oklab` a default without breaking backwards compat.

**Comment by Loirooriol** [comment]
Can you post a screenshot of the demo to see the better results? I tried WebKitGTK and `in oklab` parses but I don't see the difference.

**Comment by sesse** [comment]
> I guess that must be temporary, otherwise it would break feature detection! @sesse could clarify.

I pulled out of CSS Color a long time ago, so I don't have anything to do with Chrome's implementation, sorry. (I ended up doing nesting instead.)

**Comment by LeaVerou** [comment]
> Can you post a screenshot of the demo to see the better results? I tried WebKitGTK and `in oklab` parses but I don't see the difference.

I just added some to the first post.

**Comment by svgeesus** [comment]
So where are we on this? I saw @fserb saying that Chrome won't change gradients that currently interpolate in sRGB because backwards-compat is prioritized over better results.

**Comment by LeaVerou** [comment]
My understanding is that Chrome at some point tried interpolating in linear RGB by default and decided the backwards compat cost was too much; but that's very different than interpolating in OkLab…

Where is that comment by @fserb ?

**Comment by svgeesus** [comment]
@fserb [wrote](https://github.com/w3c/csswg-drafts/issues/7907#issuecomment-1289093556) (in the context of which interpolation spaces should be allowed and whether to add intent-based names which might change over time):

> I'm not convinced that we could do this: r**ight now, we are not changing the default behavior from legacy-srgb, because that would break content. If we can't change gradients without breaking content,** then having "intent-names" would be bad, because we would have the same problem when trying to change them in the future. I think this forces us out of intent and into explicit opt-ins.

**Comment by svgeesus** [comment]
> My understanding is that Chrome at some point tried interpolating in linear RGB by default and decided the backwards compat cost was too much; but that's very different than interpolating in OkLab…

Here is a [three-way comparison: srg-linear, srgb, and oklab](https://codepen.io/svgeesus/pen/KKBGmXG)

**Comment by svgeesus** [comment]
@fserb could yo confirm whether your backwards-compat concern was "we can't change to `oklab` to interpolate legacy" and not "we can't change to `srgb-linear` to interpolate legacy" because, as @LeaVerou pointed out, those are very different. I agree that interpolating in `srgb-linear` (which would be tempting if your colors are stored like that internally) would be a non-starter because it is very perceptually non-linear. While `srgb` and `oklab` are much more comparable, with `srgb` tending to be darker and a bit less saturated at the midpoint.
![comparison](https://user-images.githubusercontent.com/2506926/217331422-1ca1aaa5-785b-436e-837a-22127e7efdc9.png)

**Comment by svgeesus** [comment]
[Here you go](https://codepen.io/svgeesus/pen/xxJevOx) simple edit

![image](https://user-images.githubusercontent.com/2506926/217333052-572e5f87-1628-497c-9a49-0c1b1697c74c.png)

**Comment by romainmenke** [comment]
Thank you, much easier to judge the outcome :

<img width="1470" alt="Screenshot 2023-02-07 at 19 27 48" src="https://user-images.githubusercontent.com/11521496/217333565-07ff864f-045d-4d1e-8951-e821a99e9507.png">

**Comment by kizu** [comment]
Question: would it be possible to introduce a new CSS property that would set the default interpolation for the colors? Like, `interpolation-color-space` or something?

This way it would be possible to not break backwards-compat, but anyone who would want to just always use the more modern version, could just put this property in their reset. Something like `:root { interpolation-color-space: oklab; }` if it would be inherited, or `* { … }` otherwise?

**Comment by LeaVerou** [comment]
> Question: would it be possible to introduce a new CSS property that would set the default interpolation for the colors? Like, `interpolation-color-space` or something?
> 
> This way it would be possible to not break backwards-compat, but anyone who would want to just always use the more modern version, could just put this property in their reset. Something like `:root { interpolation-color-space: oklab; }` if it would be inherited, or `* { … }` otherwise?

It's not that simple, because you typically want a different interpolation color space depending on the use case (e.g. gradients vs transitions vs compositing etc).
I believe there is another issue where we discuss this, but I can't find it now.

**Comment by astearns** [comment]
> I believe there is another issue where we discuss this, but I can't find it now.

Perhaps https://github.com/w3c/csswg-drafts/issues/7035?

**Comment by romainmenke** [comment]
I still think this is a bad idea.
Please reconsider this change.

The benefits and downsides haven't been weighted properly in my opinion.

It has only been demonstrated that `oklab` is a better interpolation color space.
It hasn't been demonstrated that changing this would be a good thing.

Reasons not to do this :
- what if a future color space is even better than `oklab`? do we change to that? If so, why should users ever write a gradient without an explicit interpolation color space. They would opt-in to future breakage by omitting the color space.
- authors already have plenty of tools to reduce typing and make it easier to always opt-in to a specific interpolation color space. We don't need to make a breaking change to help authors.
- this change is known to affect existing pages, please be more careful with backwards compat.

**Comment by fserb** [comment]
Sorry, for the delay in replying. I was away. 

I agree with @romainmenke. This is not about the gradients being better looking (they are), but as mentioned before, they can have unexpected side effects on contrast and other elements close to the gradient.

Here's an example that I made using the codepen above:
<img width="1702" alt="image" src="https://user-images.githubusercontent.com/1587/227300682-ff400767-a352-41d7-bd29-ea7f59041bd2.png">

It's a simple `#F00` to `#0F0` gradient with the text color set to `#CFA903`. This is clearly readable on the current default and unreadable on oklab. This is one of many potential types of problems we will have if we change the default.

The "opt-in" change is relatively simple (just add `in oklab` on the gradient) and much safer from breaking existing sites.

**Comment by LeaVerou** [comment]
@fserb This example is specifically crafted for this purpose, but how prominent is this problem in the wild? Also do note that even in the first example, contrast is already insufficient. 

As I pointed out in the call, there is precedent for us making a change with similar dangers, when we moved from device RGB to sRGB. I don't recall anyone having text readability concerns then, and arguably that was an even more impactful change (since it also affected spot colors, not just transitions between them).

**Comment by romainmenke** [comment]
> As I pointed out in the call, there is precedent for us making a change with similar dangers, when we moved from device RGB to sRGB. I don't recall anyone having text readability concerns then, and arguably that was an even more impactful change (since it also affected spot colors, not just transitions between them).

Correct me if I am wrong, but the change from device RGB to sRGB was homogenous across any colors defined in CSS right?

That change also made CSS defined colors more consistent and predictable overal.
It made it easier to work with color on the web.

This change however doesn't do any of that.
It makes gradients more visually pleasing but authors can already opt-in to the those without breaking backwards compat.

I don't think these two changes are actually similar in nature.

-------

> This example is specifically crafted for this purpose, but how prominent is this problem in the wild

The same is true for the inverse.
Any examples in favor of this change are crafted to show visually pleasing gradients out of context. But it is the context that is the important part.

It hasn't been demonstrated that authors want this change.

**Comment by romainmenke** [comment]
@LeaVerou 

> I’m not sure what future you are reacting so strongly against.

I didn't think I was reacting overly strong here.
Written communication for non-native english speakers can lack nuance.
If I didn't express myself correctly, then I apologize, this was not my intention.

> I will remind you that we have a WG resolution and I personally will not engage in this further.

I don't think it is helpful or welcoming if the slightest friction escalates like this.
I hope we can keep this conversation open to feedback?

------

My only concern here is that backwards compat is given sufficient weight here given the small gains of this change.  Authors only need to type two words to opt-in.

To me it felt as if backwards compat wasn't given enough weight.
But as that the next step is to research exactly that, then I think there is no issue here :)

**Comment by romainmenke** [comment]
@svgeesus Was there a resolution to change this, or was there a resolution to investigate change this?

I assumed the second from the minutes, but now I am unsure.

**Comment by svgeesus** [comment]
> The final resolution in these minutes was to change the specification.

Yes. Which amounted to commenting-out a "may interpolate in sRGB' and flipping a paragraph about how to get the old behavior.

**Comment by svgeesus** [comment]
I see that there is a WPT which tests the old behavior [(legacy syntax colors interpolate in sRGB)](https://github.com/web-platform-tests/wpt/blob/master/css/css-images/gradient/legacy-color-gradient.html)

  https://github.com/web-platform-tests/wpt/blob/master/css/css-images/gradient/legacy-color-gradient.html

I guess that should be reworded and the reference should be linked with a "must not match".

@mysteryDate you wrote that test (which was correct, at the time) do you agree?

**Comment by mysteryDate** [comment]
> I guess that should be reworded and the reference should be linked with a "must not match".

Or we could just change the reference for that test to `oklab-gradient-ref.html`, because now the gradient _must_ be in oklab, right?

**Comment by svgeesus** [comment]
> Or we could just change the reference for that test to `oklab-gradient-ref.html`, because now the gradient _must_ be in oklab, right?

Yes, right.

**Comment by Gingeh** [comment]
Should http://wpt.live/css/css-images/parsing/gradient-interpolation-method-valid.html be updated to reflect this change?

**Comment by yisibl** [comment]
For the impact of changing the default to oklab will definitely be more widespread, we need to gather further feedback from the industry.

In particular, the color selector in mainstream design software is basically based on sRGB.

**Comment by majaha** [comment]
I agree with @svgeesus, the test should be updated. At the moment, the repo is in an inconsistent state, and has been for over a year now. This is pretty confusing when it's the spec prose and [MDN](https://developer.mozilla.org/en-US/docs/Web/CSS/color-interpolation-method) VS. the spec test and actual browser behaviour. I was about to file an issue about the spec not addressing legacy colour values properly until I came across this issue.

I made a jsfiddle so you can see how your browser interpolates in different situations:
https://jsfiddle.net/y7deoc2g/6/

**Comment by svgeesus** [comment]
So, over a year on, where are we?

- WPT still tests for gradients interpolating in `sRGB` by default
- no browser has changed to `oklab` by default, either in production or in an opti-in testing mode
   - so we have no data about real-world acceptability and breakage
 - the spec has said `oklab` by default for 18 months
   - browsers newly implementing this have to reverse-engineer "what browsers really do"

Agenda+ to re-examine the resolution

**Comment by tabatkins** [comment]
If the WPTs have remained testing the old, now incorrect behavior, then I'm not surprised that browsers haven't followed the spec. Those need fixing.

**Comment by Gingeh** [comment]
Despite being a pain to implement, I would personally be in favour of keeping the existing sRGB + oklab behaviour. This is primarily motivated by a gradient I found in a preexisting test page which appears much more "washed-out" when automatically changed from srgb to oklab:
![Image](https://github.com/user-attachments/assets/aa2d595f-0d44-4d07-a8a4-5af0c783f221)

**Comment by svgeesus** [comment]
So we (still) have two options

1. Change WPT to do what the spec says, assume not many sites break, hope browsers change
2. Change the spec back to what it used to say, thus aligning with WPT and current implementations

**Comment by svgeesus** [comment]
> 2\. Change the spec back to what it used to say, thus aligning with WPT and current implementations

This option amounts to reverting [this edit](https://github.com/w3c/csswg-drafts/commit/cc85d46b321003cc3849bd028ed5280a4e71c82b) and note that, cautiously, I made the edit using commenting-out rather than deletion.

If we go that way, I would argue for **not** restoring the previous text that started with:

> However, user agents &lt;em>may&lt;/em> handle interpolation between legacy sRGB color formats

because otherwise the spec is less testable (and also because no user agent did that).

**Comment by svgeesus** [comment]
[Spec updated](https://github.com/w3c/csswg-drafts/commit/a0a988606379f01917446ede6b89e83002fc148a)

**Comment by svgeesus** [comment]
The spec is updated: [12. Color Interpolation](https://drafts.csswg.org/css-color-4/#interpolation), the WPT and the spec text are in alignment, and the WPT results are pretty good.
